src/Sulu/Bundle/SecurityBundle/Factory/UserFactory.php

```diff
@@ -0,0 +1,152 @@
+        UserRepositoryInterface $userRepository,
+        RoleRepositoryInterface $roleRepository,
+        ContactRepositoryInterface $contactRepository,
+        LocalizationManagerInterface $localizationManager,
+        SaltGenerator $saltGenerator,
+        EncoderFactoryInterface $encoderFactory
+    ) {
+        $this->entityManager = $entityManager;
+        $this->userRepository = $userRepository;
+        $this->roleRepository = $roleRepository;
+        $this->contactRepository = $contactRepository;
+        $this->localizationManager = $localizationManager;
+        $this->saltGenerator = $saltGenerator;
+        $this->encoderFactory = $encoderFactory;
+    }
+
+    public function create(
+        string $username,
+        string $firstName,
+        string $lastName,
+        string $email,
+        string $locale,
+        string $password,
+        string $roleName = 'User'
+    ): UserInterface {
+        $existing = $this->userRepository->findOneBy(['username' => $username]);
+        if ($existing instanceof UserInterface) {
+            return $existing;
+        }
+
+        $locales = \array_keys($this->localizationManager->getLocalizations());
+
+        return $this->createUser($firstName, $lastName, $email, $username, $password, $locale, $roleName, $locales);
+    }
+
+    /**
+     * Generates a random salt for the password.
+     */
+    private function generateSalt(): string
+    {
+        return $this->saltGenerator->getRandomSalt();
+    }
+
+    /**
+     * Encodes the given password, for the given password, with he given salt and returns the result.
+     *
+     * @param string $user
+     * @param string $password
+     * @param string $salt
+     */
+    private function encodePassword($user, $password, $salt): string
+    {
+        return $this->encoderFactory->getEncoder($user)->encodePassword($password, $salt);
+    }
+
+    private function createUser(
+        string $firstName,
+        string $lastName,
+        string $email,
+        string $username,
+        string $password,
+        string $locale,
+        string $roleName,
+        array $locales
+    ): UserInterface {
+        $user = $this->userRepository->createNew();
+
+        /** @var ContactInterface $contact */
+        $contact = $this->contactRepository->createNew();
+        $contact->setFirstName($firstName);
+        $contact->setLastName($lastName);
+        $contact->setMainEmail($email);
+
+        $this->entityManager->persist($contact);
+        $this->entityManager->flush();
+
+        $user->setContact($contact);
+        $user->setUsername($username);
+        $user->setSalt($this->generateSalt());
+        $user->setPassword($this->encodePassword($user, $password, $user->getSalt()));
+        $user->setLocale($locale);
+        $user->setEmail($email);
+
+        /** @var RoleInterface|null $role */
+        $role = $this->roleRepository->findOneBy(['name' => $roleName]);
+        Assert::notNull($role);
```

Comment:
`Webmozart/Assert` is not part of our `composer.json`, therefore it is not safe to use it here. Is it necessary at all? Doesn't `findOneBy` throw an exception if no row is returned at all?

Reply:
No, unfortunately `findOneBy` does not throw an exception.